### PR TITLE
Adding hardswish, opinfo tests to custom rules

### PR DESCRIPTION
**Pull request by Gamrix**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69399

Differential Revision: [D32937576](https://our.internmc.facebook.com/intern/diff/D32937576)

### CODE REVIEW

File: test/jit/test_dtype_analysis.py [hunk 1]
```diff
@@ -36,11 +99,22 @@ def tearDown(self):
 
     @staticmethod
     def node_output_dtype(graph):
```
Comment (Krovatkin) [inline test/jit/test_dtype_analysis.py:101]: nit: dtype **s**?

File: test/jit/test_dtype_analysis.py [hunk 2]
```diff
@@ -245,3 +320,87 @@ def func(input, weight, bias, y):
...
+        self.prop_dtype_on_graph(graph, input_tensors)
+        self.assert_output_dtype_equal(expected_res, graph)
+
+    @ops([op for op in op_db if op.aten_name in custom_rules_works_list])
+    def test_custom_rules_ints(self, device, dtype, op):
+        # This is done because opinfos currently only runs on floats.
+        # Return fn, inputs_fn for all
+        sample_input = op.sample_inputs(device, dtype, requires_grad=False)[0]
+
+        if dtype == torch.float32:
```
Comment (Krovatkin) [inline test/jit/test_dtype_analysis.py:364]: how is `dtype` used? Should this code come before `sample_input = op.sample_inputs(device, dtype, requires_grad=False)[0]` ?

File: test/jit/test_dtype_analysis.py [hunk 3]
```diff
@@ -245,3 +320,87 @@ def func(input, weight, bias, y):
...
+                self.assert_tensor_dtype_equal(expected, actual)
+
+    def assert_tensor_dtype_equal(self, tensor_output, graph_dtype):
+        if not isinstance(tensor_output, torch.Tensor):
+            return
+        self.assertEqual(tensor_output.dtype, graph_dtype)
+
+    @ops([op for op in op_db if op.aten_name in custom_rules_works_list])
+    def test_custom_rules(self, device, dtype, op):
+        # Return fn, inputs_fn for all
```
Comment (Krovatkin) [inline test/jit/test_dtype_analysis.py:342]: nit: `test_custom_rules_ints` and `test_custom_rules` look very similar. maybe in a follow-up PR we could move out the common code into a helper.

File: test/jit/test_dtype_analysis.py [hunk 4]
```diff
@@ -19,8 +31,59 @@
     )
 
 
-# XXX: Relies on Symbolic Shape Analysis, which is still a prototype
-class TestDtypeAnalysis(JitTestCase):
+custom_rules_works_list = {
```
Comment (Krovatkin) [inline test/jit/test_dtype_analysis.py:34]: `custom_rules_works_list` and `custom_rules_works_list` should cover all custom rules we borrowed from the old shape analysis

File: torch/csrc/jit/passes/utils/op_registry.cpp [hunk 5]
```diff
@@ -52,6 +52,10 @@ std::shared_ptr<OperatorSet> nn_ops_first_input_preserving() {
       "aten::upsample_nearest3d(Tensor self, int[] output_size, float? scales_d, float? scales_h, float? scales_w) -> Tensor",
       "aten::upsample_trilinear3d(Tensor self, int[] output_size, bool align_corners, float? scales_d, float? scales_h, float? scales_w) -> Tensor",
       "aten::prelu(Tensor self, Tensor weight) -> Tensor",
+
+      // Added because Hardswish is really hard to convert to metatensors
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/utils/op_registry.cpp:56]: we shouldn't have done it much earlier :((((( spent so much time on `hardswish` :(

File: test/jit/test_dtype_analysis.py [hunk 6]
```diff
@@ -245,3 +320,87 @@ def func(input, weight, bias, y):
...
+        self.assertEqual(tensor_output.dtype, graph_dtype)
+
+    @ops([op for op in op_db if op.aten_name in custom_rules_works_list])
+    def test_custom_rules(self, device, dtype, op):
+        # Return fn, inputs_fn for all
+        sample_input = op.sample_inputs(device, dtype, requires_grad=False)[0]
+        input_args = [sample_input.input, *sample_input.args]
+        expected_res = op(*input_args, **sample_input.kwargs)
+        func = op.get_op()
+        traced_fn = create_traced_fn(self, func)
```
Comment (Krovatkin) [inline test/jit/test_dtype_analysis.py:347]: out of curiosity, why not scripting?

File: test/jit/test_dtype_analysis.py [hunk 7]
```diff
@@ -245,3 +320,87 @@ def func(input, weight, bias, y):
...
+        # Have to run the traced function to actually generate the trace
+        traced_fn(sample_input.input, *sample_input.args, **sample_input.kwargs)
+
+        # Run the Dtype Analysis
+        graph = traced_fn.graph  # Note this is a cached graph
+        input_tensors = [t for t in input_args if isinstance(t, torch.Tensor)]
+        self.prop_dtype_on_graph(graph, input_tensors)
+        self.assert_output_dtype_equal(expected_res, graph)
+
+    @expectedFailure
```
Comment (Krovatkin) [inline test/jit/test_dtype_analysis.py:383]: lmao, this is subtle, does it take a regex, so we could narrow down to our asserts?

**Review by Krovatkin — COMMENTED** [review]
Looks gut, a few small questions to make sure I understand the tests

**Review by Krovatkin — APPROVED** [review]
actually, I don't think my questions should be blocking landing
:shipit:

File: test/jit/test_dtype_analysis.py (hunk 4 above)
Comment (Gamrix) [reply test/jit/test_dtype_analysis.py:34]: We are still relying on shape analysis for shape analysis (for knowing when we have zero-dim tensors)

File: torch/csrc/jit/passes/utils/op_registry.cpp (hunk 5 above)
Comment (Gamrix) [reply torch/csrc/jit/passes/utils/op_registry.cpp:56]: Yes, totally.

File: test/jit/test_dtype_analysis.py (hunk 6 above)
Comment (Gamrix) [reply test/jit/test_dtype_analysis.py:347]: @eellison I remember that you recommended Tracing here. Could you elaborate here?

Comment (eellison) [reply test/jit/test_dtype_analysis.py:347]: Tracing more reliably works than scripting at getting IR

**Comment by Gamrix** [comment]
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32937576).
